# Task 1: Data Preprocessing and removing duplicate values from Dataset


### AIM : To remove duplicate values from the dataset.

## About the Dataset

The data used in this notebook is "Airbnb Open Data". 

### Background

It is a dataset about Airbnb, Inc., which is an American company based in San Francisco and founded in 2008, that operates an online marketplace for short- and long-term homestays and experiences. The company acts as a broker and charges a commission from each booking.

### Content

The dataset is comprehensive which contains listings on Airbnb in New York, details of the hosts, the locations, property types, prices, service fees, reviews and others.

***

## Importing Necessary Libraries

It is important for us to import necessary libraries as below.

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

***

## Data Preprocessing

First, we will be loading the dataset before performing data preprocessing.

In [5]:
data = pd.read_csv('../Desktop/Airbnb_Open_Data.csv',low_memory=False)
data.head()

id                                              NAME      host id  \
0  1001254                Clean & quiet apt home by the park  80014485718   
1  1002102                             Skylit Midtown Castle  52335172823   
2  1002403               THE VILLAGE OF HARLEM....NEW YORK !  78829239556   
3  1002755                                               NaN  85098326012   
4  1003689  Entire Apt: Spacious Studio/Loft by central park  92037596077   

  host_identity_verified host name neighbourhood group neighbourhood  \
0            unconfirmed  Madaline            Brooklyn    Kensington   
1               verified     Jenna           Manhattan       Midtown   
2                    NaN     Elise           Manhattan        Harlem   
3            unconfirmed     Garry            Brooklyn  Clinton Hill   
4               verified    Lyndon           Manhattan   East Harlem   

        lat      long        country  ... service fee minimum nights  \
0  40.64749 -73.97237  United States  ...       $193            10.0   
1  40.75362 -73.98377  United States  ...        $28            30.0   
2  40.80902 -73.94190  United States  ...       $124             3.0   
3  40.68514 -73.95976  United States  ...        $74            30.0   
4  40.79851 -73.94399  United States  ...        $41            10.0   

  number of reviews last review  reviews per month review rate number  \
0               9.0  10/19/2021               0.21                4.0   
1              45.0   5/21/2022               0.38                4.0   
2               0.0         NaN                NaN                5.0   
3             270.0    7/5/2019               4.64                4.0   
4               9.0  11/19/2018               0.10                3.0   

  calculated host listings count  availability 365  \
0                            6.0             286.0   
1                            2.0             228.0   
2                            1.0             352.0   
3                            1.0             322.0   
4                            1.0             289.0   

                                         house_rules license  
0  Clean up and treat the home the way you'd like...     NaN  
1  Pet friendly but please confirm with me if the...     NaN  
2  I encourage you to use my kitchen, cooking and...     NaN  
3                                                NaN     NaN  
4  Please no smoking in the house, porch or on th...     NaN  

[5 rows x 26 columns]

Now, we wish to explore the number of rows and columns of the data.

In [3]:
data.shape

(102599, 26)

There are 102599 rows and 26 columns in the dataset. Next, we want to check for missing values or invalid data in the dataset.

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 26 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              102599 non-null  int64  
 1   NAME                            102349 non-null  object 
 2   host id                         102599 non-null  int64  
 3   host_identity_verified          102310 non-null  object 
 4   host name                       102193 non-null  object 
 5   neighbourhood group             102570 non-null  object 
 6   neighbourhood                   102583 non-null  object 
 7   lat                             102591 non-null  float64
 8   long                            102591 non-null  float64
 9   country                         102067 non-null  object 
 10  country code                    102468 non-null  object 
 11  instant_bookable                102494 non-null  object 
 12  cancellation_pol

The output above lists all the columns in the dataset, the non-null count and the type of data for each columns. We can check the number of null values using `df.isnull().sum()` function.

In [5]:
data.isnull().sum()

id                                     0
NAME                                 250
host id                                0
host_identity_verified               289
host name                            406
neighbourhood group                   29
neighbourhood                         16
lat                                    8
long                                   8
country                              532
country code                         131
instant_bookable                     105
cancellation_policy                   76
room type                              0
Construction year                    214
price                                247
service fee                          273
minimum nights                       409
number of reviews                    183
last review                        15893
reviews per month                  15879
review rate number                   326
calculated host listings count       319
availability 365                     448
house_rules     

There are four columns with more than 10000 missing values, namely `last review`, `reviews per month`, `house_rules` and `license`. We decide to drop `last review`, `reviews per month` and`license` since we will not need it in the analysis and the number of missing values is too much. 

We will also drop `country` and `country code` since there is just one country in this dataset, namely United States. Since we have `host id`, we can drop `host name` since it serves the same purpose.

Also, null values in the `house_rules` columns are filled with `"Not stated."`. Then, we drop the rows which contains any null values in them.

In [6]:
data.drop(columns=["host name","license","last review","reviews per month","country","country code"], axis = 1, inplace = True)
data["house_rules"].fillna("Not stated.", inplace = True)
data.dropna(axis = 0, how = "any", inplace = True, ignore_index = True)
data.shape

(99886, 20)

After dropping the rows with null values, there are 99886 rows left. We would also like to drop any duplicated rows which are present in the dataset by using `drop_duplicates()` function.

In [7]:
data.drop_duplicates(keep = 'first', inplace = True, ignore_index = True)
data.shape

(99361, 20)

There are 99361 rows remaining, which we have successfully dropped 525 duplicated rows in the dataset!

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99361 entries, 0 to 99360
Data columns (total 20 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              99361 non-null  int64  
 1   NAME                            99361 non-null  object 
 2   host id                         99361 non-null  int64  
 3   host_identity_verified          99361 non-null  object 
 4   neighbourhood group             99361 non-null  object 
 5   neighbourhood                   99361 non-null  object 
 6   lat                             99361 non-null  float64
 7   long                            99361 non-null  float64
 8   instant_bookable                99361 non-null  object 
 9   cancellation_policy             99361 non-null  object 
 10  room type                       99361 non-null  object 
 11  Construction year               99361 non-null  float64
 12  price                           

Upon further investigation, the Dtype for price and service fee should be a numeric type instead of object. The reason is because the entries in these two columns are in the form of `$X,XXX` which is a string. We will fix the formats of these columns using the following codes. 

Also, the construction year, minimum nights, number of reviews, review rate number, calculated host listings count and availability 365  should be integer instead of float.

Next, instant_bookable column should be a boolean type since there are just two possible values which are True or False.

In [9]:
price=data['price'].str[1:] # remove the symbol $
price=price.str.strip() # trim whitespace

for i in range(len(price)):
    price[i]=price[i].replace(",","") # remove the symbol ,
    
svc=data['service fee'].str[1:]
svc=svc.str.strip()

for i in range(len(svc)):
    svc[i]=svc[i].replace(",","")

In [10]:
data.drop(columns=["price", "service fee"], axis = 1, inplace = True)
data["price"] = price
data["service fee"] = svc

data["price"] = pd.to_numeric(data["price"])
data["service fee"] = pd.to_numeric(data["service fee"])

data["Construction year"] = data["Construction year"].astype(int)
data["minimum nights"] = data["minimum nights"].astype(int)
data["number of reviews"] = data["number of reviews"].astype(int)
data["review rate number"] = data["review rate number"].astype(int)
data["calculated host listings count"] = data["calculated host listings count"].astype(int)
data["availability 365"] = data["availability 365"].astype(int)

data["instant_bookable"] = data["instant_bookable"].astype(bool)

In [11]:
data["neighbourhood group"].value_counts().sort_index()

neighbourhood group
Bronx             2615
Brooklyn         40543
Manhattan        42402
Queens           12873
Staten Island      927
brookln              1
Name: count, dtype: int64

Notice that there is one typo present in the column of neighbourhood group, where Brooklyn is misspelled as brookln. This typo will be fixed.

In [12]:
data.replace(to_replace='brookln', value='Brooklyn',inplace=True)

Also, the minimum nights (minimum number of night stay for the listing) must be greater than 0. We also want to focus on listings with a shorter period of stay, we limit the minimum nights to 365 (which is still considered a long period of stay). We will filter the dataset to remove any illogical samples.

In [13]:
data = data[(data["minimum nights"]>0) & (data["minimum nights"]<=365)]

In [14]:
data.head()

id                                              NAME      host id  \
0  1001254                Clean & quiet apt home by the park  80014485718   
1  1002102                             Skylit Midtown Castle  52335172823   
2  1003689  Entire Apt: Spacious Studio/Loft by central park  92037596077   
3  1004098         Large Cozy 1 BR Apartment In Midtown East  45498551794   
4  1005202                                   BlissArtsSpace!  90821839709   

  host_identity_verified neighbourhood group       neighbourhood       lat  \
0            unconfirmed            Brooklyn          Kensington  40.64749   
1               verified           Manhattan             Midtown  40.75362   
2               verified           Manhattan         East Harlem  40.79851   
3               verified           Manhattan         Murray Hill  40.74767   
4            unconfirmed            Brooklyn  Bedford-Stuyvesant  40.68688   

       long  instant_bookable cancellation_policy        room type  \
0 -73.97237             False              strict     Private room   
1 -73.98377             False            moderate  Entire home/apt   
2 -73.94399             False            moderate  Entire home/apt   
3 -73.97500              True            flexible  Entire home/apt   
4 -73.95596             False            moderate     Private room   

   Construction year  minimum nights  number of reviews  review rate number  \
0               2020              10                  9                   4   
1               2007              30                 45                   4   
2               2009              10                  9                   3   
3               2013               3                 74                   3   
4               2009              45                 49                   5   

   calculated host listings count  availability 365  \
0                               6               286   
1                               2               228   
2                               1               289   
3                               1               374   
4                               1               219   

                                         house_rules  price  service fee  
0  Clean up and treat the home the way you'd like...    966          193  
1  Pet friendly but please confirm with me if the...    142           28  
2  Please no smoking in the house, porch or on th...    204           41  
3                  No smoking, please, and no drugs.    577          115  
4  House Guidelines for our BnB We are delighted ...   1060          212

Now, the data is properly cleaned, and duplicate rows are removed.

Removing duplicate values from the dataset enhances data quality, ensuring accuracy and reliability. This process streamlines data analysis, promotes better decision-making, and establishes a foundation for deriving meaningful insights.